### ISSUE TITLE
lint/goroutineInLoop: warn on loop variables use in goroutines without copying values

### ISSUE THREAD

**Issue by NexoMichael** [go-critic/go-critic#339]
Name: goroutineInLoop
Before:

```golang
for i := range [10]int{} {
	go func() {
		fmt.Println(i) // will not print all possible values of "i" only once
	}()
}
```
After:

```golang
for i := range [10]int{} {
	func(i int) {
		defer f(i) 
	}(i)
}
```

**Comment by cristaloleg** [comment]
Name: `shadowVarInGoLoop` ? (or any better name  ;D )

**Comment by cristaloleg** [comment]
Do we want warn about any local variable in such usage or only from range(`for k, v := range ...`) ?

**Comment by NexoMichael** [comment]
More ideas for name:
* referencedVarInGoLoop
* anonymousFuncInGoLoop

**Comment by NexoMichael** [comment]
@cristaloleg good question, I think that variables defined in loop definition - are 100% issue when used in goroutines without copying, but other variables - it can be normal to have some "shared" values that are not changed in goroutines for example by convention, what also could be checked that variables are not mutable for example (but it is already race detector work to do that...)

**Comment by cristaloleg** [comment]
That's what I wanted to clarify. So any local variable inside loop's body is ok from my side:
```go
for i := range ... {
    foo := bar
    j := i
    go func() {
        _ = foo
        _ = bar
        _ = j
        _ = i // warn
    }()
}
```

**Comment by Quasilyte** [comment]
God of terse names gets nothing today.

**Comment by cristaloleg** [comment]
AFAIR it's covered by `go vet`, isn't it?

**Comment by Antonboom** [comment]
Covered by https://github.com/golangci/golangci-lint/issues/1404#issuecomment-1742002567

**Comment by cristaloleg** [comment]
Thanks to https://github.com/golang/go/wiki/LoopvarExperiment we don't need to do that (+ `go vet`).

**Comment by cristaloleg** [comment]
Thanks to https://github.com/golang/go/wiki/LoopvarExperiment we don't need to do that (+ `go vet`).
